osu.Game/Rulesets/Objects/SliderEventGenerator.cs

```diff
@@ -0,0 +1,128 @@
+
+                        var pathProgress = d / length;
+                        var timeProgress = reversed ? 1 - pathProgress : pathProgress;
+
+                        yield return new SliderEventDescriptor
+                        {
+                            Type = SliderEventType.Tick,
+                            SpanIndex = span,
+                            SpanStartTime = spanStartTime,
+                            StartTime = spanStartTime + timeProgress * spanDuration,
+                            PathProgress = pathProgress,
+                        };
+                    }
+
+                    if (span < spanCount - 1)
+                    {
+                        yield return new SliderEventDescriptor
+                        {
+                            Type = SliderEventType.Repeat,
+                            SpanIndex = span,
+                            SpanStartTime = startTime + span * spanDuration,
+                            StartTime = spanStartTime + spanDuration,
+                            PathProgress = (span + 1) % 2,
+                        };
+                    }
+                }
+            }
+
+            double totalDuration = spanCount * spanDuration;
+
+            // Okay, I'll level with you. I made a mistake. It was 2007.
+            // Times were simpler. osu! was but in its infancy and sliders were a new concept.
+            // A hack was made, which has unfortunately lived through until this day.
+            //
+            // This legacy tick is used for some calculations and judgements where audio output is not required.
+            // Generally we are keeping this around just for difficulty compatibility.
+            // Optimistically we do not want to ever use this for anything user-facing going forwards.
+
+            int finalSpanIndex = spanCount - 1;
+            double finalSpanStartTime = startTime + finalSpanIndex * spanDuration;
+            double finalSpanTime = Math.Max(startTime + totalDuration / 2, (finalSpanStartTime + spanDuration) - (legacyLastTickOffset ?? 0));
+            double finalProgress = (finalSpanTime - finalSpanStartTime) / spanDuration;
+            if (spanCount % 2 == 0) finalProgress = 1 - finalProgress;
+
+            yield return new SliderEventDescriptor
+            {
+                Type = SliderEventType.LegacyLastTick,
+                SpanIndex = finalSpanIndex,
+                SpanStartTime = finalSpanStartTime,
+                StartTime = finalSpanTime,
+                PathProgress = finalProgress,
+            };
+
+            yield return new SliderEventDescriptor
+            {
+                Type = SliderEventType.Tail,
+                SpanIndex = spanCount - 1,
+                SpanStartTime = startTime + (spanCount - 1) * spanDuration,
+                StartTime = startTime + totalDuration,
+                PathProgress = spanCount % 2,
+            };
+        }
+    }
+
+    public struct SliderEventDescriptor
```

Comment:
Members in this struct should be xmldoc'd. Interested in `SpanIndex`, `SpanStartTime`, and `PathProgress` specifically (in the context of repeats).